cpp/src/lists/contains.cu

```diff
@@ -46,6 +48,62 @@ auto get_search_keys_device_iterable_view(cudf::scalar const& search_key, rmm::c
   return &search_key;
 }
 
+template <typename ElementType, duplicate_find_option find_option>
+auto __device__ find_begin(list_device_view const& list)
+{
+  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
+    return list.pair_rep_begin<ElementType>();
+  } else {
+    return thrust::make_reverse_iterator(list.pair_rep_end<ElementType>());
+  }
+}
+
+template <typename ElementType, duplicate_find_option find_option>
+auto __device__ find_end(list_device_view const& list)
+{
+  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
+    return list.pair_rep_end<ElementType>();
+  } else {
+    return thrust::make_reverse_iterator(list.pair_rep_begin<ElementType>());
+  }
+}
+
+template <duplicate_find_option find_option, typename Iterator>
+size_type __device__ distance([[maybe_unused]] Iterator begin, Iterator end, Iterator find_iter)
+{
+  if (find_iter == end) {
+    // Not found.
+    return absent_index;
+  }
+
+  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
+    // Distance of find_position from begin.
+    return find_iter - begin;
+  } else {
+    // Distance of find_position from end.
+    return end - find_iter - 1;
+  }
```

Comment:
My preference is for single line comments to go after the line (this is totally just style preference - feel free to ignore)
```suggestion
  if (find_iter == end) {
    return absent_index;  // Not found
  }

  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
    return find_iter - begin;  // Distance of find_position from begin
  } else {
    return end - find_iter - 1;  // Distance of find_position from end
  }
```

Reply:
Agreed. This reads better.